Firmware update in command mode

lsubash opened the pull request:
This patch adds generic functionality to process Flash descriptor lock. It follows
Capsule Firmware update flow and interface is updated. Command (CMDI) interface is added
to GenCapsuleFirmware which takes file with command as input.

Sample Command format in text file input,
{FLASHDESCLOCK}

Firmware update lib handler parses high level commands
Specific command process and functionlity would be
performed by platform specific libraries.

Signed-off-by: Subash Lakkimsetti <subash.lakkimsetti@intel.com>

sagraw2 commented on PayloadPkg/FirmwareUpdate/CmdFwUpdate.c:
```diff
@@ -0,0 +1,161 @@
+    return Status;
+  }
+
+  DEBUG((DEBUG_ERROR, "Commands Requested = 0x%x\n", CmdData.CmdRequested));
```
DEBUG_INFO ?

sagraw2 commented on PayloadPkg/FirmwareUpdate/CmdFwUpdate.c:
```diff
@@ -0,0 +1,161 @@
+  DEBUG((DEBUG_ERROR, "Commands Requested = 0x%x\n", CmdData.CmdRequested));
+
+  if(CmdData.CmdRequested & CMDI_TYPE_SPI_DESCRIPTOR_LOCK) {
+    DEBUG((DEBUG_ERROR, "FLASH DESCRIPTOR LOCK COMMAND RECEIVED!! \n"));
```
Skip this trace message.

sagraw2 commented on PayloadPkg/FirmwareUpdate/CmdFwUpdate.c:
```diff
@@ -0,0 +1,161 @@
+
+  while ((CurrLine != NULL) && (CurrLine < (Buffer + BufLength))) {
+    NextLine = GetNextLine (CurrLine, &LineLen);
+    EndLine  = CurrLine + LineLen;
```
Are we using EndLine variable?

lsubash replied:
EndLine is not utilized. I would remove it.

sagraw2 commented on PayloadPkg/FirmwareUpdate/CmdFwUpdate.c:
```diff
@@ -0,0 +1,161 @@
+  @retval  EFI_NOT_FOUND    No matching payload found in the capsule.
+**/
+EFI_STATUS
+FindImage (
```
Is this forward declaration?

lsubash replied:
Yes.

gdong1 commented on PayloadPkg/FirmwareUpdate/FirmwareUpdate.c:
```diff
@@ -893,6 +894,9 @@ ApplyFwImage (
     break;
+  case FW_UPDATE_COMP_CMD_UPDATE:
+    Status = FwCmdUpdateProcess (CapImage, CapImageSize);
+    break;
```
should we call FwCmdUpdateProcess () with parameter ImageHdr to avoid search that image again?

lsubash replied:
Sure Guo. I would check this.

mauricema commented on BootloaderCommonPkg/Include/Library/FirmwareUpdateLib.h:
```diff
@@ -44,6 +44,8 @@ SPDX-License-Identifier: BSD-2-Clause-Patent
 #define FW_UPDATE_COMP_BIOS_REGION SIGNATURE_32('B', 'I', 'O', 'S')
 #define FW_UPDATE_COMP_CSME_REGION SIGNATURE_32('C', 'S', 'M', 'E')
 #define FW_UPDATE_COMP_CSME_DRIVER SIGNATURE_32('C', 'S', 'M', 'D')
+#define FW_UPDATE_COMP_CMD_UPDATE   SIGNATURE_32('C', 'M', 'D', 'I')
```
Change to "FW_UPDATE_COMP_CMD_REQUEST"

lsubash replied:
I will change the definition.

mauricema commented on BootloaderCorePkg/Tools/GenCapsuleFirmware.py:
```diff
@@ -448,6 +448,8 @@ def main():
         'MISC': '66030B7A-47D1-4958-B73D-00B44B9DD4B6',  # SBL  COMPONENT
         'CSME': '43AEF186-0CA5-4230-B1BD-193FB4627201',  # IFWI ME/CSME
         'CSMD': '4A467997-A909-4678-910C-E0FE1C9056EA',  # CSME Update driver
+        'CMDI': '9034cab1-4d19-4856-a3cd-f074263c4a4d',  # FLEX Command
```
# Command Request

mauricema commented on PayloadPkg/FirmwareUpdate/CmdFwUpdate.c:
```diff
@@ -0,0 +1,161 @@
+
+  DEBUG((DEBUG_ERROR, "Commands Requested = 0x%x\n", CmdData.CmdRequested));
+
+  if(CmdData.CmdRequested & CMDI_TYPE_SPI_DESCRIPTOR_LOCK) {
```
Is this common for all platform?  If not, this needs to be moved into platform FWU library for command handling.

lsubash replied:
Yes, interface should be same for all platforms.
Parser would populate CmdData.CmdRequested. If this command is in generic, it would be executed from common. If its platform specific (current FlashDescLock), platform FWU library would handle it.

mauricema commented on Silicon/ApollolakePkg/Library/FirmwareUpdateLib/FirmwareUpdateLib.c:
```diff
@@ -495,18 +495,26 @@ PlatformEndFirmwareUpdate (
 EFIAPI
-ClearFwUpdateTrigger (
-  VOID
+SetFlashDescriptorLock (
```
If it is a common interface, it should be
FwUpdateCmdHandler (Args)

lsubash replied:
For APL, I have removed ClearFwUpdateTrigger () by mistake. I would fix it.

lsubash commented on PayloadPkg/FirmwareUpdate/CmdFwUpdate.c:
```diff
@@ -0,0 +1,161 @@
+  CHAR8        *FlashDescLockDataBuf;
+  // Flash Descriptor Lock command Data length
+  UINT32      FlashDescLockDataLen;
+} CMDI_DATA;
```
Currently this data structure is limited to FlashDesc. Same structure can be appended as below.
Parser would update these fields and respective subfunction would invoked next.

typedef struct {
  // Commands requested
  CMDI_TYPE   CmdRequested;
  // Flash Descriptor Lock command Data
  CHAR8        *FlashDescLockDataBuf;
  // Flash Descriptor Lock command Data length
  UINT32      FlashDescLockDataLen;

  **CHAR8        *USecase2DataBuf;
  UINT32      USecase2DataBufLen;
  CHAR8        *USecase3DataBuf;
  UINT32      USecase3DataBufLen;**

} CMDI_DATA;

mauricema replied:
My thought was that this function will only split the text file into individual command and arguments.  Then it call the command handler in a chain to process the command.  At top level, we don't parse what is the detailed command at all.

lsubash replied:
Update FwCmdHandler to process the request.
Error handling need to improved in case of multiple command request, Currently we use only one command.

lsubash commented:
Updated PR,
-  FwCmdUpdateProcess () with parameter ImageHdr
- Fixed comments provided
- Fixed issues reported by automated build triggered with this patch

gdong1 commented on BootloaderCommonPkg/Include/Library/FirmwareUpdateLib.h:
```diff
@@ -503,4 +505,41 @@ EFIAPI
+EFI_STATUS
+FwCmdUpdateProcess (
+  EFI_FW_MGMT_CAP_IMAGE_HEADER  *CapImageHdr
+  );
```
Library/FirmwareUpdateLib.h is the interfaces file provided by FirmwareUpdateLib.
FwCmdUpdateProcess () is not part of FirmwareUpdateLib, should not be added to this header file.
it could be added to a internal header file in Firmware update driver (FirmwareUpdateHelper.h?)

lsubash replied:
Done. Moved FwCmdUpdateProcess() to local header FirmwareUpdateHelper.h.

mauricema commented on PayloadPkg/FirmwareUpdate/CmdFwUpdate.c:
```diff
@@ -0,0 +1,146 @@
+  @retval  other            error status from the update routine
+**/
+EFI_STATUS
+FwCmdUpdateProcess (
```
I think this function can be fully moved into platform library.  In this way, we don' t need to add new platform handler for each new command. 
I am OK if we can handle it as a 2nd patch.
